### PR TITLE
multi-account support for s3 buckets

### PR DESCRIPTION
**Pull request by jaymccon**
## Overview

enables s3_sync to work with multiple accounts/buckets.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -24,6 +24,6 @@ coverage.xml
 /.pytest_cache/
 /.mypy_cache/
 taskcat_outputs/
-
+foo.py
```
Comment (jaymccon) [inline .gitignore]: What's this for?

File: taskcat/legacy_cli.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
+#!/usr/bin/env python
```
Comment (jaymccon) [inline taskcat/legacy_cli.py]: Is this file the result of a bad merge?

**Comment by jaymccon** [comment]
Just a few comments from an initial scan of the pr, will put in thorough review on Monday

**Comment by jaymccon** [comment]
still needs jsonschema additions to work:

```
conf = Config(project_config_path="./tests/data/nested_fail/ci/taskcat.yml", project_root="./tests/data/nested_fail/")
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat-v9/.venv/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-e4edc1cd7d8d>", line 1, in <module>
    conf = Config(project_config_path=c, project_root=p)
  File "/Users/jmmccon/taskcat-v9/taskcat/_config.py", line 102, in __init__
    self._process_project_config()
  File "/Users/jmmccon/taskcat-v9/taskcat/_config.py", line 339, in _process_project_config
    validate(instance, "project_config")
  File "/Users/jmmccon/taskcat-v9/taskcat/_common_utils.py", line 174, in schema_validate
    resolver=RefResolver(str(schema_path.as_uri()) + "/", None),
  File "/Users/jmmccon/taskcat-v9/.venv/lib/python3.7/site-packages/jsonschema/validators.py", line 541, in validate
    cls(schema, *args, **kwargs).validate(instance)
  File "/Users/jmmccon/taskcat-v9/.venv/lib/python3.7/site-packages/jsonschema/validators.py", line 130, in validate
    raise error
jsonschema.exceptions.ValidationError: Additional properties are not allowed ('boto_profile' was unexpected)
Failed validating 'additionalProperties' in schema['properties']['general']:
    {'$id': '#general',
     'additionalProperties': False,
     'properties': {'auth': {'$ref': '#/definitions/auth'},
                    'aws_access_key': {'type': 'string'},
                    'aws_secret_key': {'type': 'string'},
                    'config_file': {'type': 'string'},
                    'enable_sig_v2': {'type': 'boolean'},
                    'exclude': {'type': 'string'},
                    'lambda_build_only': {'type': 'boolean'},
                    'lint': {'type': 'boolean'},
                    'no_cleanup': {'type': 'boolean'},
                    'no_cleanup_failed': {'type': 'boolean'},
                    'parameters': {'$ref': '#/definitions/parameters'},
                    'profile_name': {'type': 'string'},
                    'project_root': {'type': 'string'},
                    'public_s3_bucket': {'type': 'boolean'},
                    'stack_prefix': {'type': 'string'},
                    'tags': {'$ref': '#/definitions/tags'},
                    'upload_only': {'type': 'boolean'},
                    'verbosity': {'enum': ['debug',
                                           'info',
                                           'warning',
                                           'error'],
                                  'type': 'string'}},
     'type': 'object'}
On instance['general']:
    {'boto_profile': 'test'}
```

**Comment by jaymccon** [comment]
there's also a merge conflict on `taskcat/_config.py` that needs to be addressed

**Comment by jaymccon** [comment]
Would be nice to have a complex `taskcat.yml` file that puts this through it's paces, not sure what format we landed on, but something like:

```
general:
  boto_profile: default
tests:
  supplies_own_default:
    boto_profile: notdefault
  use_default_except_ue1:
    regions:
      - us-east-1: 
        boto_profile: someotherprofile
      - us-west-2
      - eu-central-1
```

Will also be useful as a base test case once tests are added.

File: tests/test_cfn_lint.py [hunk 3]
```diff
@@ -2,11 +2,26 @@
 import unittest
 from pathlib import Path
 
+import mock
```
Comment (jaymccon) [inline tests/test_cfn_lint.py]: none of this mocking is needed, just have to add `create_clients=False` to the test's `Config` instantiation

**Comment by jaymccon** [comment]
@andrew-glenn from my tests the functional tests still not, well, functional.

can you please address and comment with examples of some functional tests that you performed.

```python
from pathlib import Path
from taskcat import Config
c = Config(
    project_config_path=Path('./tests/data/regional_client_and_bucket/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/regional_client_and_bucket/').resolve()
)
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat-v9/.venv/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-410895617135>", line 4, in <module>
    project_root=Path('./tests/data/regional_client_and_bucket/').resolve()
  File "/Users/jmmccon/taskcat-v9/taskcat/_config.py", line 111, in __init__
    self._process_project_config()
  File "/Users/jmmccon/taskcat-v9/taskcat/_config.py", line 352, in _process_project_config
    instance["tests"][test], project_root=self.project_root
  File "/Users/jmmccon/taskcat-v9/taskcat/_config_types.py", line 109, in from_dict
    return Test(**raw_test)
  File "/Users/jmmccon/taskcat-v9/taskcat/_config_types.py", line 28, in __init__
    self.template_file: Path = self._guess_path(template_file)
  File "/Users/jmmccon/taskcat-v9/taskcat/_config_types.py", line 51, in _guess_path
    f"Cannot find {path} with project root" f" {self._project_root}"
taskcat.exceptions.TaskCatException: Cannot find debug-json.template with project root /Users/jmmccon/taskcat-v9/tests/data/regional_client_and_bucket
```

File: taskcat/_config.py [hunk 4]
```diff
@@ -131,6 +146,89 @@ def _get_templates(self):
...
+            client_factory = self._client_factory_instance.return_credset_instance(
+                cred_key
+            )
+            if client_factory:
+                region.client = client_factory
+                region.client.set = True
+                sts_client = region.client.get("sts")
+                try:
+                    account = sts_client.get_caller_identity()["Account"]
+                except Exception:
```
[the remark this answers is not in the document]
Comment (andrew-glenn) [reply taskcat/_config.py]: Will fix.

File: tests/test_cfn_lint.py (hunk 3 above)
Comment (andrew-glenn) [reply tests/test_cfn_lint.py]: Right. this is now outdated. thanks.

File: taskcat/_config.py [hunk 5]
```diff
@@ -131,6 +146,89 @@ def _get_templates(self):
...
+        if name in bucket_dict.keys():
+            return bucket_dict[name]
+        bucket_instance = S3Bucket(name=name, **kwargs)
+        if name:
+            bucket_dict[name] = bucket_instance
+        if account:
+            bucket_dict[account] = bucket_instance
+        return bucket_instance
+
+    def _assign_account_buckets(self):
```
[the remark this answers is not in the document]
Comment (andrew-glenn) [reply taskcat/_config.py]: discussed offline; no longer a factor.

File: taskcat/_config.py [hunk 6]
```diff
@@ -131,6 +147,92 @@ def _get_templates(self):
...
+                cred_key
+            )
+            if not client_factory:
+                continue
+            region.client = client_factory
+            region.client.set = True
+            sts_client = region.client.get("sts")
+            try:
+                account = sts_client.get_caller_identity()["Account"]
+            except ClientError as e:
```
Comment (jaymccon) [inline taskcat/_config.py]: can we also trap `botocore.exceptions.NoCredentialsError: Unable to locate credentials` I ran into this by accident as I had disabled my default creds

File: taskcat/_config.py [hunk 7]
```diff
@@ -131,6 +147,92 @@ def _get_templates(self):
...
+        ]
+        for cred_key in cred_key_list:
+            client_factory = self._client_factory_instance.return_credset_instance(
+                cred_key
+            )
+            if not client_factory:
+                continue
+            region.client = client_factory
+            region.client.set = True
+            sts_client = region.client.get("sts")
```
Comment (jaymccon) [inline taskcat/_config.py]: try and except here for botocore.exceptions.ProfileNotFound: The config profile (<PROFILE_NAME>) could not be found will allow the cli to handle unmatched profiles gracefully

**Comment by andrew-glenn** [comment]
@jaymccon  - The changes we discussed yesterday are now committed.

File: tests/test_cfn_lint.py (hunk 3 above)
Comment (jaymccon) [reply tests/test_cfn_lint.py]: don't see that this has been addressed

File: taskcat/_config.py (hunk 7 above)
Comment (jaymccon) [reply taskcat/_config.py]: ?

File: taskcat/_config.py (hunk 6 above)
Comment (jaymccon) [reply taskcat/_config.py]: ?

File: tests/test_cfn_lint.py (hunk 3 above)
Comment (andrew-glenn) [reply tests/test_cfn_lint.py]: This is actually necessary.

File: taskcat/_config_types.py [hunk 8]
```diff
@@ -109,12 +109,39 @@ def from_dict(cls, raw_test: dict, project_root="./"):
...
+    def __init__(self, region_name: str, client_factory: ClientFactory):
+        self.name: str = region_name
+        self.s3bucket = None
+        self.account: Optional[str] = None
+        self._cf: ClientFactory = client_factory
+        self._credset_name: Optional[str] = None
+        self._credset_modify = True
+
+    @property
+    def credset_name(self) -> Optional[str]:
```
Comment (jaymccon) [inline taskcat/_config_types.py]: wouldn't calling this profile_name be more intuitive ?

File: taskcat/_config_types.py [hunk 9]
```diff
@@ -109,12 +109,39 @@ def from_dict(cls, raw_test: dict, project_root="./"):
...
+        return self._credset_name
+
+    @credset_name.setter
+    def credset_name(self, value: str) -> None:
+        if not self._credset_modify:
+            return
```
Comment (jaymccon) [inline taskcat/_config_types.py]: should this just silently do nothing? Might be better to raise an exception here.

File: taskcat/_template_params.py [hunk 10]
```diff
@@ -156,7 +156,7 @@ def get_content(self, bucket, object_key):
         :return: Content of the object
 
         """
-        s3_client = self._boto_client.get("s3", region=self.region, s3v4=True)
+        s3_client = self._boto_client("s3", s3v4=True)
```
Comment (jaymccon) [inline taskcat/_template_params.py]: no longer need the `s3v4=True` as it's the default

File: taskcat/_template_params.py [hunk 11]
```diff
@@ -341,9 +341,7 @@ def _get_presigned_url_wrapper(self, presigned_url_regex):
                     url_bucket, url_key, url_expire_seconds
                 )
             )
-            s3_client = self._boto_client.get(
-                "s3", region=self._boto_client.get_default_region(), s3v4=True
-            )
+            s3_client = self._boto_client("s3", s3v4=True)
```
Comment (jaymccon) [inline taskcat/_template_params.py]: no longer need the s3v4=True as it's the default

File: taskcat/_config_types.py (hunk 8 above)
Comment (andrew-glenn) [reply taskcat/_config_types.py]: Sure.

Comment (andrew-glenn) [reply taskcat/_config_types.py]: Actuallllllllly. No.  This could be any of these...

https://github.com/taskcat/taskcat-v9/blob/9a57ec02961019d08d1e773a5ecc9078b46bfe62/taskcat/_config.py#L155-L160

File: taskcat/_config_types.py (hunk 9 above)
Comment (andrew-glenn) [reply taskcat/_config_types.py]: Yes. The intention is to iterate through a list of regions, set the credential_set appropriately, then "lock" the changes. Example:

https://github.com/taskcat/taskcat-v9/blob/9a57ec02961019d08d1e773a5ecc9078b46bfe62/taskcat/_config.py#L161-L180

File: taskcat/_config.py [hunk 12]
```diff
@@ -374,3 +476,131 @@ def _harvest_env_vars(self, env_vars):
                 elif value.lower() in ["true", "false"]:
                     value = value.lower() == "true"
                 self.env_vars[key] = value
+
+
+class S3Bucket:
```
Comment (jaymccon) [inline taskcat/_config.py]: in the interest of keeping this file small (well, to be fair, lets call it `less humungous`) can we move this out into it's own file.

File: taskcat/_config_types.py [hunk 13]
```diff
@@ -109,12 +109,39 @@ def from_dict(cls, raw_test: dict, project_root="./"):
...
-class S3BucketConfig(str):
-    def __init__(self, public: bool = False, auto: bool = False):
-        self.region = ""
-        self.public = public
-        self.auto = auto
-        self.max_name_len = 63
-        self.name = self.__str__()
-        self.tags: list = []
-        super().__init__()
+class S3BucketConfig:
```
Comment (jaymccon) [inline taskcat/_config_types.py]: this could share a file with `S3Bucket`

**Comment by andrew-glenn** [comment]
Example of `<region>.client` behavior.

```
>>> t = c.tests['yaml-test']
>>> region = t.regions[0]
>>> ec2 = region.client('ec2')
>>>
>>> import pprint as pp
>>> pp.pprint(ec2.describe_availability_zones())
{'AvailabilityZones': [{'Messages': [],
                        'RegionName': 'ap-northeast-1',
                        'State': 'available',
                        'ZoneId': 'apne1-az4',
                        'ZoneName': 'ap-northeast-1a'},
                       {'Messages': [],
                        'RegionName': 'ap-northeast-1',
                        'State': 'available',
                        'ZoneId': 'apne1-az1',
                        'ZoneName': 'ap-northeast-1c'},
                       {'Messages': [],
                        'RegionName': 'ap-northeast-1',
                        'State': 'available',
                        'ZoneId': 'apne1-az2',
                        'ZoneName': 'ap-northeast-1d'}],
 'ResponseMetadata': {'HTTPHeaders': {'content-length': '1005',
                                      'content-type': 'text/xml;charset=UTF-8',
                                      'date': 'Thu, 01 Aug 2019 22:26:11 GMT',
                                      'server': 'AmazonEC2'},
                      'HTTPStatusCode': 200,
                      'RequestId': '<REDACTED>',
                      'RetryAttempts': 0}}
>>>
```

File: tests/data/regional_client_and_bucket/ci/taskcat.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    - us-east-2
+    - us-west-1
+    - us-west-2
+    - sa-east-1
+  auth:
+    us-east-1: prod
+    sa-east-1: south-america-prod
+    us-east-2: staging
+tests:
+#  json-test:
```
Comment (jaymccon) [inline tests/data/regional_client_and_bucket/ci/taskcat.yml]: If we're not going to use this test can we remove it and it's associated files from the pr

Comment (jaymccon) [reply tests/data/regional_client_and_bucket/ci/taskcat.yml]: on 2nd thoughts I think we need this to be re-enabled. without it we're not looking at a broad enough set of possible overrides.

**Comment by jaymccon** [comment]
This does not seem to be working as expected.

using this to create config:
```python
from pathlib import Path
from taskcat import Config
c = Config(
    project_config_path=Path('./tests/data/regional_client_and_bucket/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/regional_client_and_bucket/').resolve()
)
```

If I run the above without a `foobar` profile setup in my `~/.aws/credentials` I expect to have it raise an exception due to missing profile. Instead shat happens is that is completes init successfully and all regions are set to default:

```
[r._credset_name for r in c.tests['yaml-test'].regions]
Out[6]: 
['default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default',
 'default']
```

Same results if I add the foobar profile. what I would expect is that the `sa-east-1` region uses the default profile, and all other regions use `foobar`.

Here are my test criteria for approving this pr:

* profile should be configurably overridable 
* the override behavior should be intuitive 
* work as expected in all edge cases
* The s3bucket name should be generated to be unique per account
* where different profiles map to the same account a single bucket should be shared between profiles

**Comment by andrew-glenn** [comment]
@jaymccon 

Functional test code...
```
print("Test-based region verification")
for test_name, test_obj in c.tests.items():
    print(test_name)
    for region in test_obj.regions:
        print(f"\tRegion: {region.name}\tCredset Name: {region._credset_name}\tAccount: {region.account}\tS3Bucket: {region.s3bucket.name}")
```

Functional test output.
```
Test-based  region          verification
yaml-test
Region:     ap-northeast-1  Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     ap-northeast-2  Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     ap-south-1      Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     ap-southeast-1  Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     ap-southeast-2  Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     ca-central-1    Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     eu-central-1    Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     eu-west-1       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     eu-west-2       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     us-east-1       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     us-east-2       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     us-west-1       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     us-west-2       Credset       Name:  yaml-test_default    Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     sa-east-1       Credset       Name:  yaml-test_sa-east-1  Account:  XXXXXXX61708  S3Bucket:  taskcat-complex-auth-config-file-45bcad46
Region:     me-south-1      Credset       Name:  me-south-1           Account:  XXXXXXX54109  S3Bucket:  taskcat-complex-auth-config-file-f3f86a4b
Region:     ap-east-1       Credset       Name:  ap-east-1            Account:  XXXXXXX31334  S3Bucket:  taskcat-complex-auth-config-file-adc27ecb
```

**Comment by jaymccon** [comment]
LGTM

Only concern is that instantiating the config object is really slow (10sec+ in my tests) let's address that in a future PR

**Comment by jaymccon** [comment]
@andrew-glenn can you approve, seeing as I opened the PR I cannot approve it :/
